### ISSUE TITLE
Supporting ARM64 CPUs on systems without /proc/cpuinfo

### ISSUE THREAD

**Issue by VariantXYZ** [sonos/tract#852]
Related to #847 

> It is possible to override the detector with environment variables... It's not documented, I thought it was only making sens in qemu test contexts where detection is confused. See https://github.com/sonos/tract/blob/main/linalg/src/arm64.rs#L61/

(I just happened to hit this issue since /proc/cpuinfo didn't exist)

> Which part of what getrandom does do you refer to ? The ability to register a generator at runtime ?

Right, the ability to provide a fallback implementation for unsupported targets. 

https://docs.rs/getrandom/latest/getrandom/macro.register_custom_getrandom.html

I'm not a huge fan of it though, as it involves modifying the crate to add the fallback implementation for a platform. 

The environment variable setup would work, but it would be more ideal to have this available during compile-time. 

I was thinking about using [features](https://doc.rust-lang.org/reference/attributes/codegen.html#the-target_feature-attribute), but ARM is a bit strange such that in-order/out-of-order execution is actually implied by the CPU name, which is a bit of a pain. Things like fp16, neon, etc... can be handled via this though (with the caveat that target_feature can only be applied to unsafe functions). Maybe the CPU ID could be provided via a regular feature?

In a situation where a user is building on a platform that isn't windows/linux, I think it's not unreasonable to expect them to provide a .cargo/config.toml or something similar to define their setup better.

**Comment by kali** [comment]
tract strategy for cpu features is to build everything and use runtime detection to pick the right kernel just in time. If it is possible, I would prefer to keep it like that. 

For runtime switch, I am pretty happy with the stdlib feature detection on x64. For arm64, [is_aarch64_feature_detected](https://doc.rust-lang.org/std/arch/macro.is_aarch64_feature_detected.html#) is available starting with rust 1.60, which is higher than tract MSRV at this stage (I could bump it pretty soon, 1.60 is already six month old). I will be happy to get rid of my cpuinfo parser :)

On top of feature detection, as you noticed, we need the CPU part for the Cortex variants. Do you happen to know how the feature detection is actually implemented on FreeBSD by the stdlib runtime? With no /proc/cpuinfo and no permission to run cpuid userland, I guess there is some OS trickery somewhere. Is it working at least? If it does it should point us towards the FreeBSD way to query the OS for CPU parts or some kind identifier info we could use to get what we need.

Just out of curiosity: what actual platorm are you running on? FreeBSD on Apple Silicon? (You don't have to tell if you don't want)

**Comment by VariantXYZ** [comment]
> tract strategy for cpu features is to build everything and use runtime detection to pick the right kernel just in time. If it is possible, I would prefer to keep it like that.

Noted. I assumed tract's goal might've been to simplify targeting specific platforms, so I figured compile-time definitions would be ideal, but it's not a big deal. The overhead isn't generally noticeable and is_aarch64_feature_detected should be plenty.

I suppose a minor disadvantage would be some system register access isn't exposed (e.g. it's locked to EL0 and some registers are only available in EL1), but this is probably a rare case that shouldn't come up.

> Do you happen to know how the feature detection is actually implemented on FreeBSD by the stdlib runtime?

sysctl exists, but I assume for aarch64, most information can be pulled with system register information (at least, that's what I gather looking at [std::detect](https://github.com/rust-lang/stdarch/tree/master/crates/std_detect/src/detect/os)).

Unfortunately I'm not sure of a better way to get the exact CPU model in a 'standard' way for FreeBSD... In this case, I think the correct thing to do would be to manually override or allow for specifying it during compile time.

> Just out of curiosity: what actual platorm are you running on? FreeBSD on Apple Silicon? (You don't have to tell if you don't want)

Hah, I apologize, it's nothing so exciting. I found an old BeagleBone with FreeBSD running a Cortex-A8 and wanted to check if it could do some ML inference.

**Comment by kali** [comment]
Ok, I think we have a solution in sight:

1/ bump tract MSRV at the first occasion (in next batch of breaking changes)
2/ use stdlib feature detection (mostly for fp16 activation)
3/ use cpu part from /proc/cpuinfo where possible, with a fallback to the generic implementation
4/ keep the environment variable override alive

Is this correct / acceptable ?

**Comment by VariantXYZ** [comment]
> 4/ keep the environment variable override alive

This works, though I think having some way to provide an implementation for a CPU (and maybe an example for adding one) would be helpful in the future.

**Comment by VariantXYZ** [comment]
@kali ,

> This works, though I think having some way to provide an implementation for a CPU (and maybe an example for adding one) would be helpful in the future.

Actually, I tried building for cortex-a57 with the neon feature enabled (just happened to be the first out-of-order CPU that seemed close enough to cortex-a53/a55) and noticed that the generic 4x4 kernel for mmm does not do any SIMD operations (which would be a huge slowdown). My assumption is that it's because Rust's bounds checks are preventing LLVM from auto-vectorizing the generic implementation (BLAS implementations in C++ like OpenBLAS tend to have no issue auto-vectorizing).

I'm not sure what the best path forward is to get the best simd-optimized implementations for a generic aarch64 target, but I think that mechanism would be useful. Maybe this is out of scope for the current ticket?

**Comment by kali** [comment]
This is completely out-of-scope :)

I spent some time trying to make auto-vectorization work, but never managed it in Rust. There are some issues with fast math that are highly controversial (for good reasons). As long as Rust does not have a story there (a safe one), autovectorization of this kind of code can not work.

Anyway, I don't trust autovectorizer (and to some extent intrinsics-based code) and prefer to bypass entirely the compiler around the critical topic of matrix multiplication. The generic Rust-written kernel are ultimate fallbacks (and reference implementation for tests). For Cortex-a57, tract should select the arm64simd_mmm_f32_*_generic family. It will probably  not be as good as a hand crafted a57 targetting kernel that does not exist, but it will be way better than the Rust fallbacks.

**Comment by VariantXYZ** [comment]
> For Cortex-a57, tract should select the arm64simd_mmm_f32_*_generic family. 

Funny enough, after I posted that comment, I remembered I had a homebrewed Switch lying around (cortex-a57, woo) running Ubuntu and decided to running some samples. Building aarch64-unknown-linux-gnu with target-cpu=cortex-a57 and target-feature=+neon. A gprof of running resnet showed that it was spending some time in the generic 4x4 function. I suppose the fact that it's not falling back on arm64simd is a bug?

Out of curiosity, I decided to force the KIND to cortex-a53, just to see if it'd change implementations, and it did. Seemed to be noticeably faster too.

I can open another ticket for this, as it seems like this may be going a bit off topic.
